### ISSUE TITLE
Timeout - main consumer queue is full

### ISSUE THREAD

**Issue by guybluebat** [python-beaver/python-beaver#391]
I'm experiencing a problem with a number of beaver instances where, after operating without issue for a number of hours, I start seeing the following error continuously in the log:

```
INFO    Connected
ERROR   Queue is full
ERROR   Recieved timeout from main consumer queue - stopping queue
INFO    Starting queue consumer
```

This repeats endlessly. If I restart beaver, it reloads and starts to operate properly. 

When this started, I was using beaver v34.1.0, since then I upgraded to 36.0.1 and finally 36.1.0 but it still seems to be an issue.

I should note that I'm running beaver inside a docker container and accessing logs from the host and from other containers as volumes, so that could be a cause of this issue.

**Comment by guybluebat** [comment]
As a test I removed the docker container setup and instead installed beaver directly on my machines, leaving it overnight. When I checked it again the next day I found that the same error had shown up on at least one of the servers running beaver.

**Comment by imacube** [comment]
The error "Queue is Full" means beaver's internal queue is full of log messages to send and won't continue reading the file until messages in the queue are sent. You would only loose log messages if the file was truncated, overwritten, or deleted (assuming beaver couldn't still use the file descriptor to access the deleted file). 

I posted about increasing the queue size in #342 .

**Comment by guybluebat** [comment]
Ok, I'll try increasing the queue size and see how it goes. Thanks for the
info.

On Thu, Feb 25, 2016 at 10:56 AM, Ryan <notifications@github.com> wrote:

> The error "Queue is Full" means beaver's internal queue is full of log
> messages to send and won't continue reading the file until messages in the
> queue are sent. You would only loose log messages if the file was
> truncated, overwritten, or deleted (assuming beaver couldn't still use the
> file descriptor to access the deleted file).
>
> I posted about increasing the queue size in #342
> <https://github.com/python-beaver/python-beaver/issues/342> .
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/python-beaver/python-beaver/issues/391#issuecomment-188929207>
> .
>



-- 

*Guy Knights*
Senior Systems Engineer
BlueBat Games Inc.
Ph: 778-379-5120
Email: guy@bluebatgames.com

**Comment by guybluebat** [comment]
I've tried increasing the queue size, first to 1000, then to 10000, but I'm still getting this issue. Can I just clarify, is the issue here that the volume of logs being processed is exceeding the ability of my output destination's ability to ingest the processed logs? Because the output I have configured is a fairly hefty AWS elasticache redis cluster, and the volume of logs I'm processing on each server is pretty low. 

My list of inputs includes apache and nginx error and access logs, php error logs, and the system log. However, there is basically no web traffic to these servers so it seems kind of unlikely that the volume of logs is exceeding the ability of redis to ingest them. I guess there could be some sort of network issue involved and I'll do some investigation on that front, but short of finding something to that effect I'm not sure what else to try.

At the moment I'm using cron to restart beaver every hour which seems to be mostly working, but it's a bandaid fix and I'd prefer to fix the root of the issue if possible.

**Comment by imacube** [comment]
I'm using beaver to send approximately 10 million logs per day via UDP to a logstash cluster. I've not worked with redis before. 

That said, I've been looking through the redis code in beaver: `beaver/transports/redis_transport.py`. The method, `redis_transport.callback`, looks up a valid redis server and connects to it every time it's called to send logs. Depending on how long its taking to find a valid redis server and connect to it, this could be eating up time such that beaver gets behind the files its monitoring--far enough behind that your queue size isn't big enough to handle the next read of log files. I would recommend setting beaver's log level to debug. This might give you a better idea of what beaver is doing--e.g. if over the hour of run time the number of logs being sent to redis in a single call to `redis_transport.callback` increases with each loop, or if beaver is having problems connecting to the redis cluster.

**Comment by guybluebat** [comment]
It's been a while, but I'm still seeing this issue on some of my servers even though I've since changed my beaver & logstash config to bypass redis altogether. Beaver now sends logs direct to logstash using the TPC transport in JSON format. However, I still get some servers getting in this state:

[2016-05-03 18:47:13,889] ERROR   Recieved timeout from main consumer queue - stopping queue
[2016-05-03 18:47:18,809] INFO    Starting queue consumer
[2016-05-03 18:47:18,889] INFO    Connected
[2016-05-03 18:47:18,889] ERROR   Queue is full

I tried running in debug mode but I didn't see much else in there to help me troubleshoot this. Is there a way I can view the contents of the queue to try and see what's going on?
